optimade/server/schemas.py

```diff
@@ -0,0 +1,118 @@
+from optimade.models import (
+    StructureResource,
+    ReferenceResource,
+    LinksResource,
+    EntryResource,
+)
+from .config import CONFIG
+
+
+def retrieve_queryable_properties(
+    schema: dict, queryable_properties: list, entry_provider_fields: list = None
+) -> dict:
+    """ For a given schema dictionary and a list of desired properties,
+    recursively descend into the schema and set the appropriate values
+    for the `description`, `sortable` and `unit` keys, returning the expanded
+    schema dict that includes nested schemas.
+
+    """
+    properties = {}
+    if entry_provider_fields is None:
+        entry_provider_fields = []
+
+    for name, value in schema["properties"].items():
+        if name in queryable_properties:
+            if "$ref" in value:
+                path = value["$ref"].split("/")[1:]
+                sub_schema = schema.copy()
+                while path:
+                    next_key = path.pop(0)
+                    sub_schema = sub_schema[next_key]
+                sub_queryable_properties = sub_schema["properties"].keys()
+                properties.update(
+                    retrieve_queryable_properties(
+                        sub_schema, sub_queryable_properties, entry_provider_fields
+                    )
+                )
+            else:
+                properties[name] = {"description": value.get("description", "")}
+                if "unit" in value:
+                    properties[name]["unit"] = value["unit"]
+                # All properties are sortable with the MongoDB backend.
+                # While the result for sorting lists may not be as expected, they are still sorted.
+                properties[name]["sortable"] = value.get("sortable", True)
+
+            if name in entry_provider_fields:
+                # Rename fields in schema according to provider field list
+                properties[f"_{CONFIG.provider.prefix}_{name}"] = properties.pop(name)
+
+    return properties
+
+
+class EntrySchemas:
+    """ Wrapper class for EntryResource schemas used in this server. Any
+    references in the schema are resolved to allow for easy use. If the
+    underyling model is updated, the schema will be refreshed.
+
+    """
+
+    def __init__(self):
+        self._structure_model = StructureResource
+        self._reference_model = ReferenceResource
+        self._links_model = LinksResource
+        self._structures = None
+        self._references = None
+        self._links = None
+
+    def keys(self):
+        return {"structures", "references"}
+
+    def get(self, key: str):
```

Comment:
`__getattr__`?

I think my question is more whether this class should properly use the built-in Python methods similar to a `dict` or it should be more mocked, like this (and `keys`)?